File: Leetcode/LinkedListCycle141.java
package Algorithms.Leetcode;

/**
 * Given a linked list, determine if it has a cycle in it.
 * Follow up:
 * Can you solve it without using extra space?
 *
 * Floyd's tortoise and the hare algorithm moves two pointers at different speeds
 * through the sequence of values until they both point to equal values.
 *
 * Created by dianaluca on 10/1/16.
 */

public class LinkedListCycle141 {
  public class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
      val = x;
      next = null;
    }
  }

  public boolean hasCycle(ListNode head) {
    if (head == null) return false;
    ListNode slow, fast;
    slow = fast = head;

    while (true) {
      slow = slow.next;

      if (fast.next == null) return false;
      else fast = fast.next.next;

      if (slow == null || fast == null) return false;

      if (slow == fast) return true;  //if the 2 nodes ever meet, means we have a loop
    }
  }
}

File: LeetcodePython/LinkedListCycle141.py
# Given a linked list, determine if it has a cycle in it.
#
# Follow up:
# Can you solve it without using extra space?

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        # initHead = head

        if not head or not head.next:
            return False

        slow = head
        fast = head.next

        while slow.val != fast.val:
            if not slow.next or not fast.next or not fast.next.next:
                return False
            slow = slow.next
            fast = fast.next.next

        return True